promremote/client.go

```diff
@@ -0,0 +1,178 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package promremote
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/m3db/m3/src/query/ts"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/prometheus/prompb"
+)
+
+const (
+	// DefaultRemoteWrite is the default Prom remote write endpoint in m3coordinator.
+	DefaultRemoteWrite = "http://localhost:7201/api/v1/prom/remote/write"
+
+	defaulHTTPClientTimeout = time.Second * 30
+)
+
+// Tag are the metric tags.
+type Tag struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+// Timeseries are made of tags and a datapoint.
+type Timeseries struct {
+	Tags      []Tag
+	Datapoint ts.Datapoint
+}
+
+type client struct {
+	writeURL   string
+	httpClient *http.Client
+}
+
+type clientOptions struct {
+	remoteWriteURL    string
+	httpClientTimeout time.Duration
+}
+
+// M3Client is used to write timeseries data to m3coordinator.
+type M3Client interface {
+	// Write writes the Prom proto WriteRequest to the specified endpoint.
+	Write(*prompb.WriteRequest) error
+}
+
+// ClientOptions defines available methods.
+type ClientOptions interface {
+	// SetWriteURL sets the URL which the client uses to write to m3coordinator.
+	SetWriteURL(string) ClientOptions
+
+	// WriteURL returns the URL which the client uses to write to m3coordinator.
+	WriteURL() string
+
+	// SetHTTPClientTimeout sets the timeout for the client.
+	SetHTTPClientTimeout(time.Duration) ClientOptions
+
+	//HTTPClientTimeout returns the timeout that is set for the client.
+	HTTPClientTimeout() time.Duration
+}
+
+// NewClientOpts returns a default clientOptions struct.
+func NewClientOpts() ClientOptions {
+	return &clientOptions{
+		remoteWriteURL:    DefaultRemoteWrite,
+		httpClientTimeout: defaulHTTPClientTimeout,
+	}
+}
+
+// NewClient creates a new remote write coordinator client.
+func NewClient(opts ClientOptions) M3Client {
+	return &client{
+		writeURL: opts.WriteURL(),
+		httpClient: &http.Client{
+			Timeout: opts.HTTPClientTimeout(),
+		},
+	}
+}
+
+func (o *clientOptions) SetWriteURL(val string) ClientOptions {
+	opts := *o
+	opts.remoteWriteURL = val
+	return &opts
+}
+
+func (o *clientOptions) WriteURL() string {
+	return o.remoteWriteURL
+}
+
+func (o *clientOptions) SetHTTPClientTimeout(val time.Duration) ClientOptions {
+	opts := *o
+	opts.httpClientTimeout = val
+	return &opts
+}
+
+func (o *clientOptions) HTTPClientTimeout() time.Duration {
+	return o.httpClientTimeout
+}
+
+func (c *client) Write(promWR *prompb.WriteRequest) error {
+	data, err := proto.Marshal(promWR)
+	if err != nil {
+		return fmt.Errorf("unable to marshal protobuf: %v", err)
+	}
+
+	encoded := snappy.Encode(nil, data)
```

Comment:
This may not be very performant if we snappy encode every point we get (won't give us any decent savings either); any chance we can have this thing batch at all? i.e. how important is it that this will not drop a single entry?

Otherwise it may be better for this to just use the JSON endpoint, or add a protobuf endpoint that takes a single point, otherwise  we're doing a ton of encoding/decoding

Reply:
This snappy encodes an entire Prom WriteRequest which can contain multiple timeseries:
```
type WriteRequest struct {
	Timeseries []*TimeSeries `protobuf:"bytes,1,rep,name=timeseries" json:"timeseries,omitempty"`
}
```